### ISSUE TITLE
label-smoothing and dataset

### ISSUE THREAD

**Issue by shining-love** [ultralytics/ultralytics#19791]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, I have two questions to ask. The first one is how to set up label-smoothing for training in the latest version of the project? The second is that I have a complex requirement. I have seen that image-weights can be used to address the issue of class imbalance, and I have also incorporated the solution provided by Y.T.G. It indeed improves the performance for classes with fewer samples, but the accuracy for classes with more samples has decreased. This might be because the classes with more samples are sampled less frequently during training. I would like to ask how to solve this problem? My idea is to first train for the first n epochs using build.YOLODataset = YOLOWeightedDataset, and then train without weighting for the remaining N-n epochs, which should lead to some improvement. However, I don't know how to write such code, so I would appreciate your help. Thank you.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @shining-love, thank you for reaching out and for your interest in Ultralytics 🚀! We appreciate your detailed questions and description.

For your inquiries:  
- If this is a ❓ Question about training adjustments like label-smoothing or dataset weighting, please provide as much information as possible, such as your training setup, relevant code snippets, and logs to help us understand your use case better.  
- If this is a 🐛 Bug Report, we kindly ask for a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug the issue.

Meanwhile, we recommend exploring the following resources, which may already have answers to your questions:
- [Docs](https://docs.ultralytics.com/) for comprehensive details on YOLO training, augmentation, and customization options.
- [Guides for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for tips on improving your training outcomes.

To ensure you're working with the latest features and fixes, update your `ultralytics` package in a [**Python>=3.8**](https://www.python.org/) environment with:
```bash
pip install -U ultralytics
```

## Join the Community
Engage with the vibrant Ultralytics community for further insights and support:
- Real-time discussions on [Discord](https://discord.com/invite/ultralytics) 🎧  
- Community posts on [Discourse](https://community.ultralytics.com/) or [Subreddit](https://reddit.com/r/Ultralytics)  

## Verified Environments
YOLO is compatible with various environments. For example:
- **Notebooks with free GPU**:  
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  
- **Cloud platforms**: See [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) or [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/) guides  
- **Docker**: Use our [Docker Image](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing, ensuring robust functionality across YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/).  

This is an automated response to help guide you. An Ultralytics engineer will review your issue and provide additional assistance soon. Thank you for your patience! 😊

**Comment by glenn-jocher** [comment]
For label smoothing in YOLO11, set the `label_smoothing` parameter in your training configuration file. See the [Training Hyperparameters](https://docs.ultralytics.com/guides/model-training-tips/#key-training-parameters) section for details.

Regarding class imbalance, dynamic dataset switching mid-training isn't natively supported. Instead, consider:
1. Adjusting class weights inversely to class frequencies
2. Using oversampling for rare classes
3. Applying balanced sampling strategies

Explore the [Dataset Preparation Guide](https://docs.ultralytics.com/guides/data-collection-and-annotation/) for implementation details. For custom approaches, you'll need to modify the dataset class directly in your training loop.

**Comment by shining-love** [comment]
> For label smoothing in YOLO11, set the `label_smoothing` parameter in your training configuration file. See the [Training Hyperparameters](https://docs.ultralytics.com/guides/model-training-tips/#key-training-parameters) section for details.
> 
> Regarding class imbalance, dynamic dataset switching mid-training isn't natively supported. Instead, consider:
> 
> 1. Adjusting class weights inversely to class frequencies
> 2. Using oversampling for rare classes
> 3. Applying balanced sampling strategies
> 
> Explore the [Dataset Preparation Guide](https://docs.ultralytics.com/guides/data-collection-and-annotation/) for implementation details. For custom approaches, you'll need to modify the dataset class directly in your training loop.

Thank you for your reply. 
（1）I see that label-smoothing has been removed in the latest version of u. So, does that mean label-smoothing cannot be set in the latest version of u?
（2）I have add image-weights method，my code as below:

`from ultralytics import YOLO
from ultralytics.data.dataset import YOLODataset
import ultralytics.data.build as build
import numpy as np
import matplotlib.pyplot as plt
import cv2
import os
from collections import Counter


class YOLOWeightedDataset(YOLODataset):
    def __init__(self, *args, mode="train", **kwargs):
        """
        Initialize the WeightedDataset.

        Args:
            class_weights (list or numpy array): A list or array of weights corresponding to each class.
        """

        super(YOLOWeightedDataset, self).__init__(*args, **kwargs)

        self.train_mode = "train" in self.prefix

        # You can also specify weights manually instead
        self.count_instances()
        class_weights = np.sum(self.counts) / self.counts
        self.agg_func = np.mean

        self.class_weights = np.array(class_weights)
        self.weights = self.calculate_weights()
        self.probabilities = self.calculate_probabilities()

    def count_instances(self):
        """
        Count the number of instances per class

        Returns:
            dict: A dict containing the counts for each class.
        """
        self.counts = [0 for i in range(len(self.data["names"]))]
        for label in self.labels:
            cls = label['cls'].reshape(-1).astype(int)
            for id in cls:
                self.counts[id] += 1

        self.counts = np.array(self.counts)
        self.counts = np.where(self.counts == 0, 1, self.counts)

    def calculate_weights(self):
        """
        Calculate the aggregated weight for each label based on class weights.

        Returns:
            list: A list of aggregated weights corresponding to each label.
        """
        weights = []
        for label in self.labels:
            cls = label['cls'].reshape(-1).astype(int)

            # Give a default weight to background class
            if cls.size == 0:
              weights.append(1)
              continue

            # Take mean of weights
            # You can change this weight aggregation function to aggregate weights differently
            # weight = np.mean(self.class_weights[cls])
            # weight = np.max(self.class_weights[cls])
            weight = self.agg_func(self.class_weights[cls])
            weights.append(weight)
        return weights

    def calculate_probabilities(self):
        """
        Calculate and store the sampling probabilities based on the weights.

        Returns:
            list: A list of sampling probabilities corresponding to each label.
        """
        total_weight = sum(self.weights)
        probabilities = [w / total_weight for w in self.weights]
        return probabilities

    def __getitem__(self, index):
        """
        Return transformed label information based on the sampled index.
        """
        # Don't use for validation
        if not self.train_mode:
            return self.transforms(self.get_image_and_label(index))
        else:
            index = np.random.choice(len(self.labels), p=self.probabilities)
            return self.transforms(self.get_image_and_label(index))



os.environ['KMP_DUPLICATE_LIB_OK']='True'

if __name__ == "__main__":
  
  # Monkey patch method,要测试这种图像加权方式是否兼容多卡训练
  build.YOLODataset = YOLOWeightedDataset
  
  """train"""
  #v12seg暂时官方还没有释放预训练模型
  model = YOLO('./ultralytics/cfg/models/attention_v11/model_atten/yolov11-seg-globalcontext.yaml').load('yolo11s-seg.pt')


  
  results = model.train(
    data='./ultralytics/cfg/datasets/my-seg.yaml',
    task="segment",
    epochs=300, 
    batch=8,
    workers=4, 
    imgsz=1024,
    scale=0.3,  
    mosaic=0.8,
    mixup=0.0,  
    copy_paste=0.2,  
    flipud=0.5,
    fliplr=0.5,
    degrees=0.3,
    translate=0.1,
    device="0",
    patience=50,
    close_mosaic=10,
    name = 'exp_v11seg',
    cache='disk',
  )
  
  
  print(model.trainer.train_loader.dataset)
  
  # Evaluate model performance on the validation set

  metrics = model.val()

  
  any other method to solve my problems as description?please help me in detail,thanks.

**Comment by glenn-jocher** [comment]
@shining-love **Answer**  

(1) Label smoothing remains available in YOLO11. You can set it via the `label_smoothing` parameter in your training configuration (e.g., `label_smoothing=0.1`). For implementation details, see the [Training Configuration documentation](https://docs.ultralytics.com/usage/cfg/#train-settings).  

(2) For class imbalance, consider these enhancements to your weighted sampling approach:  
- Implement **focal loss** to focus training on hard examples  
- Use **stratified sampling** instead of pure inverse-frequency weighting  
- Add **temperature scaling** to your class weight aggregation (`self.agg_func`)  
- Combine with **loss-based curriculum learning**  

For multi-GPU compatibility in your custom `YOLOWeightedDataset`, ensure your sampling probabilities are synchronized across processes. You might also explore the built-in `class_weights` parameter in newer YOLO11 versions.

**Comment by shining-love** [comment]
> [@shining-love](https://github.com/shining-love) **Answer**
> 
> (1) Label smoothing remains available in YOLO11. You can set it via the `label_smoothing` parameter in your training configuration (e.g., `label_smoothing=0.1`). For implementation details, see the [Training Configuration documentation](https://docs.ultralytics.com/usage/cfg/#train-settings).
> 
> (2) For class imbalance, consider these enhancements to your weighted sampling approach:
> 
> * Implement **focal loss** to focus training on hard examples
> * Use **stratified sampling** instead of pure inverse-frequency weighting
> * Add **temperature scaling** to your class weight aggregation (`self.agg_func`)
> * Combine with **loss-based curriculum learning**
> 
> For multi-GPU compatibility in your custom `YOLOWeightedDataset`, ensure your sampling probabilities are synchronized across processes. You might also explore the built-in `class_weights` parameter in newer YOLO11 versions.

https://github.com/ultralytics/ultralytics/pull/16014

This links shaw label_smoothing was removed.?

**Comment by Y-T-G** [comment]
>  I have seen that image-weights can be used to address the issue of class imbalance, and I have also incorporated the solution provided by Y.T.G. It indeed improves the performance for classes with fewer samples, but the accuracy for classes with more samples has decreased. This might be because the classes with more samples are sampled less frequently during training

Did you try training for more epochs?

**Comment by shining-love** [comment]
> > I have seen that image-weights can be used to address the issue of class imbalance, and I have also incorporated the solution provided by Y.T.G. It indeed improves the performance for classes with fewer samples, but the accuracy for classes with more samples has decreased. This might be because the classes with more samples are sampled less frequently during training
> 
> Did you try training for more epochs?
> 
> Or using different aggregation function? Like `np.sum`
> 
> https://y-t-g.github.io/tutorials/yolo-class-balancing/#is-it-perfectly-balanced

Thanks for your reply to me. I have not to try np.sum，I will try this trick("np. sum"). Gook Luck.

**Comment by glenn-jocher** [comment]
For class balancing with weighted sampling, consider adjusting the class weight calculation formula in your `YOLOWeightedDataset` class. Try using inverse class frequencies directly rather than normalized sums, and experiment with different aggregation methods like `np.sum` to maintain majority class representation. You can implement dynamic weighting by modifying the `calculate_weights()` method during training iterations. For more advanced strategies, explore the Ultralytics Class Balance Guide.

**Comment by shining-love** [comment]
> For class balancing with weighted sampling, consider adjusting the class weight calculation formula in your `YOLOWeightedDataset` class. Try using inverse class frequencies directly rather than normalized sums, and experiment with different aggregation methods like `np.sum` to maintain majority class representation. You can implement dynamic weighting by modifying the `calculate_weights()` method during training iterations. For more advanced strategies, explore the Ultralytics Class Balance Guide.

thanks for your advice on class balancing.how to using inverse class frequencies directly rather than normalized sums?could you show me some code demo?

**Comment by glenn-jocher** [comment]
@shining-love Here's a straightforward approach to using inverse class frequencies directly:

```python
# Instead of:
# class_weights = np.sum(self.counts) / self.counts

# Use inverse frequencies:
class_weights = 1.0 / self.counts  # Simple inverse

# Or with normalization to prevent extreme values:
class_weights = 1.0 / (self.counts / np.min(self.counts))  # Relative to smallest class

# You can also add smoothing to prevent extreme weights for very rare classes:
class_weights = 1.0 / (self.counts + smooth_factor)  # Where smooth_factor is something like 5 or 10
```

These approaches give higher weights to rare classes while maintaining better representation of common classes. The smoothing factor helps prevent overfitting to extremely rare classes.
